### ISSUE TITLE
`test_mnist.py` fails when x data is in [0, 1] range

### ISSUE THREAD

**Issue by gustavolsson** [geohot/tinygrad#278]
I've been playing around with tinygrad and I really like the minimalism.

The mnist pixel data is in the `[0, 255]` range and normalizing it to `[0, 1]` somehow results in much worse performance.

In `test_mnist.py` change line 18 to `return X_train / 255.0, Y_train, X_test / 255.0, Y_test` and the test will fail, _even with 10x the number of steps_. Increasing the learning-rate seem to improve the results but I can't get as good accuracy as with non-normalized inputs. What's going on here?

Is it floating point precision errors accumulating in the model, a bug in tinygrad or me not understanding the theory? (I've found some people on Stack Overflow mentioning this but they either seem to erroneously normalize the labels or not compare normalized/non-normalized inputs)

**Comment by Henistein** [comment]
Normalize the X_train is not just divide by 255, you have to find his mean and standard deviation then you use this [formula](https://miro.medium.com/max/512/1*Ap_7t_-luGSaAVgc7kl7qA.png).
`X_train = (X_train - X_train.mean()) / X_train.std()`
And do the same with X_test. But like you I am confused too about this, because I did some comparisons using just CNN model and got this (range, steps, accuracy):

- [0, 255]  200  0.97580
- [0, 1]  200  0.91850 (without normalizing)
- [0, 1]  200  0.94180 (normalizing)
So I conclude that even normalizing I can't get the same accuracy that range [0, 255] has, using the same learning rate and steps.
But this is not a problem from tinygrad but in general, I have the same issues in another models that I made. I am a beginner in this field, so if someone could explain what is happening would be great.

**Comment by Henistein** [comment]
I've been investigating a bit more about this issue and I've concluded that normalizing is generally a good practice. But in some models there may not be a significant improvement, so the best thing to do is to test several types of normalizations and choose the most appropriate one.
I did a small test that shows a small improvement when using [0, 1] normalization. Here it is:
```python
def test_ADAM(self):
    np.random.seed(1337)
    model = TinyBobNet()
    optimizer = optim.Adam(model.parameters(), lr=0.001)
    train(model, X_train, Y_train, optimizer, steps=1000)
    assert evaluate(model, X_test, Y_test) > 0.95
```
